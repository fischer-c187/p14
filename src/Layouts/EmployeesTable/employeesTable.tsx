import Table from "@components/Table/table";
import TableBody from "@components/Table/tableBody";
import TableHeader from "@components/Table/tableHeader";
import { ListEmployee } from "../../Types/api";
import employeesTableDescriptor from "../../constants/tableDescriptor";

type EmployeesTableProps = {
  data: ListEmployee;
  onColumnSort?: (sortBy: string) => void;
  sortConfig?: { sortBy: string; sortDesc: boolean };
};

function EmployeesTable({
  data,
  onColumnSort,
  sortConfig,
}: EmployeesTableProps) {
  return (
    <Table>
      <TableHeader
        tableDescriptor={employeesTableDescriptor}
        onColumnSort={onColumnSort}
        sortBy={sortConfig?.sortBy}
        sortDesc={sortConfig?.sortDesc}
      />
      <TableBody data={data} tableDescriptor={employeesTableDescriptor} />
    </Table>
  );
}

export default EmployeesTable;
