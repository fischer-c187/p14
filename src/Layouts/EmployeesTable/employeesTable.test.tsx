import { render, screen } from "@testing-library/react";
import { describe, expect, it, vi } from "vitest";
import userEvent from "@testing-library/user-event";

import EmployeesTable from "./employeesTable";

const data = [
  {
    firstName: "Alice",
    lastName: "Smith",
    dateBirth: "02-15-1985",
    startDate: "06-01-2010",
    street: "123 Oak St",
    city: "Springfield",
    state: "IL",
    zipCode: "62704",
    department: "Sales",
    _id: "0d115e8d-3f65-46b3-aba0-2ac2f3ad2f01",
  },
  {
    firstName: "Bob",
    lastName: "Johnson",
    dateBirth: "07-22-1978",
    startDate: "08-15-2012",
    street: "456 Maple St",
    city: "Riverside",
    state: "CA",
    zipCode: "92501",
    department: "Marketing",
    _id: "a0927f4b-2e17-4efc-aae2-0b9f54d994ac",
  },
  {
    firstName: "Carol",
    lastName: "Williams",
    dateBirth: "01-30-1990",
    startDate: "11-10-2014",
    street: "789 Pine St",
    city: "Greenfield",
    state: "WI",
    zipCode: "53220",
    department: "HR",
    _id: "ca0391cd-f894-4ad9-86f3-390780ef6c59",
  },
];

describe("EmployeesTable", () => {
  it("should render correctly with table header and table body", () => {
    const { asFragment } = render(<EmployeesTable data={data} />);
    screen.debug();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should call onColumnSort on click", async () => {
    const onColumnSort = vi.fn();
    const user = userEvent.setup();
    render(<EmployeesTable data={data} onColumnSort={onColumnSort} />);

    await user.click(screen.getByText("First Name"));
    screen.debug();
    expect(onColumnSort).toHaveBeenCalledWith("firstName");
  });
});
