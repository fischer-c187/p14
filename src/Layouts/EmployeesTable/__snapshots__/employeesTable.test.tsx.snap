// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmployeesTable > should render correctly with table header and table body 1`] = `
<DocumentFragment>
  <table
    class="table-auto w-full"
  >
    <thead
      class="border-b border-gray-200"
    >
      <tr
        class="bg-gray-50"
      >
        <th
          class="py-3 px-3 text-gray-500 font-semibold text-xs text-left lg:px-6 text-nowrap whitespace-nowrap cursor-pointer"
        >
          First Name
        </th>
        <th
          class="py-3 px-3 text-gray-500 font-semibold text-xs text-left lg:px-6 text-nowrap whitespace-nowrap cursor-pointer"
        >
          Last Name
        </th>
        <th
          class="py-3 px-3 text-gray-500 font-semibold text-xs text-left lg:px-6 text-nowrap whitespace-nowrap cursor-pointer"
        >
          Start Date
        </th>
        <th
          class="py-3 px-3 text-gray-500 font-semibold text-xs text-left lg:px-6 text-nowrap whitespace-nowrap cursor-pointer"
        >
          Department
        </th>
        <th
          class="py-3 px-3 text-gray-500 font-semibold text-xs text-left lg:px-6 text-nowrap whitespace-nowrap cursor-pointer"
        >
          Date of Birth
        </th>
        <th
          class="py-3 px-3 text-gray-500 font-semibold text-xs text-left lg:px-6 text-nowrap whitespace-nowrap"
        >
          Street
          <img
            alt="arrow"
            aria-hidden="true"
            class="inline-block ml-1 lg:ml-2 w-4 h-4"
            src="/src/assets/arrow-down.svg"
          />
        </th>
        <th
          class="py-3 px-3 text-gray-500 font-semibold text-xs text-left lg:px-6 text-nowrap whitespace-nowrap cursor-pointer"
        >
          City
        </th>
        <th
          class="py-3 px-3 text-gray-500 font-semibold text-xs text-left lg:px-6 text-nowrap whitespace-nowrap cursor-pointer"
        >
          State
        </th>
        <th
          class="py-3 px-3 text-gray-500 font-semibold text-xs text-left lg:px-6 text-nowrap whitespace-nowrap cursor-pointer"
        >
          Zip Code
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        class="[&>*:nth-child(-n+2)]:text-gray-900 [&>*:nth-child(-n+2)]:font-medium border border-gray-200"
      >
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          Alice
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          Smith
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          06-01-2010
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          Sales
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          02-15-1985
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          123 Oak St
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          Springfield
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          IL
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          62704
        </td>
      </tr>
      <tr
        class="[&>*:nth-child(-n+2)]:text-gray-900 [&>*:nth-child(-n+2)]:font-medium border border-gray-200"
      >
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          Bob
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          Johnson
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          08-15-2012
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          Marketing
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          07-22-1978
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          456 Maple St
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          Riverside
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          CA
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          92501
        </td>
      </tr>
      <tr
        class="[&>*:nth-child(-n+2)]:text-gray-900 [&>*:nth-child(-n+2)]:font-medium border border-gray-200"
      >
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          Carol
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          Williams
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          11-10-2014
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          HR
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          01-30-1990
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          789 Pine St
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          Greenfield
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          WI
        </td>
        <td
          class="px-3 py-4 whitespace-nowrap text-xs text-gray-500 lg:text-sm lg:px-6"
        >
          53220
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
